doc/source/scripts/gen_topologyparser_attrs.py

```diff
@@ -64,101 +65,170 @@
+            expected.add("elements")
+        parser_attrs[test_parser_class.parser] = (expected, guessed)
+    return parser_attrs
+
+
+class TopologyParsers:
+    def __init__(self) -> None:
+        def _keys(parser: TopologyReaderBase) -> tuple[str, str]:
+            f = parser.format
+            if isinstance(f, (list, tuple)):
+                key = f[0]
+                label = ", ".join(f)
+            else:
+                key = label = f
+            return (key, label)
+
+        def _description(
+            parser: TopologyReaderBase,
+            expected: set[str],
+            guessed: set[str],
+            key_label: tuple[str, str],
+        ) -> str:
+            key, label = key_label
+            return DESCRIPTIONS[key]
+
+        def _format(
+            parser: TopologyReaderBase,
+            expected: set[str],
+            guessed: set[str],
+            key_label: tuple[str, str],
+        ) -> str:
+            key, label = key_label
+            return base.sphinx_ref(txt=label, label=key, suffix="-format")
+
+        def _attributes_read(
+            parser: TopologyReaderBase,
+            expected: set[str],
+            guessed: set[str],
+            key_label: tuple[str, str],
+        ) -> str:
+            vals = sorted(expected - guessed)
+            return ", ".join(vals)
+
+        def _attributes_guessed(
+            parser: TopologyReaderBase,
+            expected: set[str],
+            guessed: set[str],
+            key_label: tuple[str, str],
+        ) -> str:
+            return ", ".join(sorted(guessed))
+
+        parser_attrs = create_parser_attributes()
+        input_items = [
+            [parser, expected, guessed, _keys(parser=parser)]
+            for parser, (expected, guessed) in parser_attrs.items()
+        ]
+        self.table_writer = TableWriter(
+            filename="formats/topology_parsers.txt",
+            include_table="Table of supported topology parsers and the attributes read",
+            sort=True,
+            input_items=input_items,
+            lines=[],
+            column_spec=[
+                ("Format", _format),
+                ("Description", _description),
+                ("Attributes read", _attributes_read),
+                ("Attributes guessed", _attributes_guessed),
+            ],
+        )
+        self.table_writer.generate_lines_and_write_table()
+
+        attrs = defaultdict(set)
+
+        def _get_attrs(
+            line: list[str],
+            format: str,
+            parser: TopologyReaderBase,
+            expected: set[str],
+            guessed: set[str],
+            key_label: tuple[str, str],
+        ) -> None:
+            for attribute in expected | guessed:
+                attrs[attribute].add(format)
+
+        for line, format, args in zip(
+            self.table_writer.lines,
+            self.table_writer.fields["Format"],
+            input_items,
+        ):
+            _get_attrs(line, format, *args)
```

Comment:
I'm not sure this code needs to remain inside the class -- TopologyParsers doesn't do anything with it as far as I can tell, and it's only used for the TopologyAttrs/ConnectivityAttrs classes below. IMO it'd be more straightforward to slightly rewrite `main` and separate this out as a function.

```suggestion
def _get_format_attrs(topology_parser: TopologyParser) -> dict[str, set[str]]:
    attrs = defaultdict(set)
    writer = topology_parser.table_writer
    for format, (expected, guessed, _) in zip(
        writer.fields["Format"],
        input_items,
    ):
        for attribute in expected | guessed:
            attrs[attribute].add(format)
    return attrs
```